src/main/java/tests/ShopServiceTest.java

```diff
@@ -0,0 +1,110 @@
+        operationHandlers.put(Operation.SUPPLY, new SupplyOperation());
+        OperationStrategy operationStrategy = new OperationStrategyImpl(operationHandlers);
+        shopService = new ShopServiceImpl(operationStrategy);
+        transactions = new ArrayList<>();
+    }
+
+    @Test
+    public void test_balance_operation() {
+        FruitTransaction balance = new FruitTransaction();
+        balance.setOperation(Operation.BALANCE);
+        balance.setFruit("banana");
+        balance.setQuantity(100);
+        transactions.add(balance);
+
+        shopService.process(transactions);
+        Map<String, Integer> storage = shopService.getStorage();
+
+        assertEquals(100, (int) storage.get("banana"));
+    }
+
+    @Test
+    public void test_supply_operation() {
+        FruitTransaction balance = new FruitTransaction();
+        balance.setOperation(Operation.BALANCE);
+        balance.setFruit("banana");
+        balance.setQuantity(100);
+        transactions.add(balance);
+
+        FruitTransaction supply = new FruitTransaction();
+        supply.setOperation(Operation.SUPPLY);
+        supply.setFruit("banana");
+        supply.setQuantity(50);
+        transactions.add(supply);
+
+        shopService.process(transactions);
+        Map<String, Integer> storage = shopService.getStorage();
+
+        assertEquals(150, (int) storage.get("banana"));
+    }
+
+    @Test
+    public void test_purchase_operation() {
+        FruitTransaction balance = new FruitTransaction();
+        balance.setOperation(Operation.BALANCE);
+        balance.setFruit("banana");
+        balance.setQuantity(100);
+        transactions.add(balance);
+
+        FruitTransaction purchase = new FruitTransaction();
+        purchase.setOperation(Operation.PURCHASE);
+        purchase.setFruit("banana");
+        purchase.setQuantity(30);
+        transactions.add(purchase);
+
+        shopService.process(transactions);
+        Map<String, Integer> storage = shopService.getStorage();
+
+        assertEquals(70, (int) storage.get("banana"));
+    }
+
+    @Test
+    public void test_combined_operations() {
+        FruitTransaction balance = new FruitTransaction();
+        balance.setOperation(Operation.BALANCE);
+        balance.setFruit("banana");
+        balance.setQuantity(100);
+        transactions.add(balance);
+
+        FruitTransaction supply = new FruitTransaction();
+        supply.setOperation(Operation.SUPPLY);
+        supply.setFruit("banana");
+        supply.setQuantity(50);
+        transactions.add(supply);
+
+        FruitTransaction purchase = new FruitTransaction();
+        purchase.setOperation(Operation.PURCHASE);
+        purchase.setFruit("banana");
+        purchase.setQuantity(30);
+        transactions.add(purchase);
+
+        shopService.process(transactions);
+        Map<String, Integer> storage = shopService.getStorage();
+
+        assertEquals(120, (int) storage.get("banana"));
```

Comment:
This test is a good start for combined operations, but you should also consider adding more complex scenarios, like handling invalid transactions in the middle of the list, to ensure the system is robust and handles such cases gracefully.